### ISSUE TITLE
Completion of Canceled Persistent Requests using MPI_TEST

### ISSUE THREAD

**Issue by omor1** [mpi-forum/mpi-issues#155]
# Problem

In MPI-3.1, it is unclear whether a single call to MPI_TEST is sufficient for completing and inactivating a canceled persistent requested such that it can be safely freed. This is the only use case where, depending on interpretation, a single call to a nonblocking completion operation can be required to complete a request.

# Proposal

Clarify that `MPI_TEST` must be called repeated in a loop to complete and inactivate such a canceled request.

# Changes to the Text

MPI-3.1 standard, page 72, lines 20–22:
>After the call to MPI_CANCEL and the subsequent call to MPI_WAIT or MPI_TEST, the request becomes inactive and can be activated for a new communication.

Should change to:
> After the call to MPI_CANCEL and the subsequent call to MPI_WAIT, the request becomes inactive and can be activated for a new communication; similarly, if MPI_TEST is repeatedly called in a busy wait loop, the request will eventually become inactive.

# Impact on Implementations

Likely none.

# Impact on Users

Users may need to use MPI_WAIT or call MPI_TEST in a loop if they do not already do so for ensuring completion of a canceled persistent request.

# References

This came about in a discussion with @bosilca of the use of persistent requests (and the cancelation thereof) in PaRSEC. The fact that there were different interpretations possible indicated that this should be brought to discussion to the Forum. Code exemplifying the problem:
```C
for( i = 0; i < am->req_count; i++ ) {
        MPI_Cancel(&am->reqs[i]);
        MPI_Test(&am->reqs[i], &flag, &status);    /* is this sufficient? */
        MPI_Request_free(&am->reqs[i]);
}
```

**Comment by dholmes-epcc-ed-ac-uk** [comment]
I agree with clarifying this. I would expect that the MPI_TEST in the code snippet would be sufficient if and only it set flag to true.

Given that, I would suggest that the wording of the change be re-worked slightly:
> After the call to MPI_CANCEL and a subsequent call to a completion function, such as a call to MPI_WAIT or a call to MPI_TEST that returns flag = true, the request becomes inactive and can be activated for a new communication.

... or ....

> After the call to MPI_CANCEL and subsequent completion of the operation, such as via a call to MPI_WAIT or a call to MPI_TEST that returns flag = true, the request becomes inactive and can be activated for a new communication.

Personally, I would want to avoid the phrase "busy-wait loop"in the standard, because the application is free to overlap computation with the repetition of calls to MPI_TEST.

**Comment by omor1** [comment]
> Personally, I would want to avoid the phrase "busy-wait loop"in the standard, because the application is free to overlap computation with the repetition of calls to MPI_TEST.

Agreed; I was reusing the language used for testing the cancelation of non-persistent requests on lines 15–17 on the same page, but your formulation is preferred.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@omor1 that suggests that this ticket should propose a wording change for the existing text that you mention to make that consistent with this new phrase here.

**Comment by omor1** [comment]
@dholmes-epcc-ed-ac-uk I'm not sure if that should be on this ticket or a new one; I don't exactly want to muddle the two issues together.

**Comment by dholmes-epcc-ed-ac-uk** [comment]
@omor1 given that text on lines 15-17, the minimal change is to add "successful" to the offending sentence.
> After the call to MPI_CANCEL and the subsequent *successful* call to MPI_WAIT or MPI_TEST, the request becomes inactive and can be activated for a new communication.

**Comment by bosilca** [comment]
A "successful call to MPI_Test" is unclear on what is expected to be successful the function call itself or the request completion. What about:
> After the call to MPI_CANCEL and the subsequent successful request completion, using any of the calls described in Section 3.7.3, the request becomes inactive and can be activated for a new communication.

**Comment by RolfRabenseifner** [comment]
Yes, George's proposal is best, but please with 3.7.5 too:

> After the call to MPI_CANCEL and the subsequent successful request completion, using any of the calls described in Sections 3.7.3 and 3.7.5, the request becomes inactive and can be activated for a new communication.
